From: Harald Freudenberger <freude@de.ibm.com>
Subject: [PATCH] [BZ 149851] kernel: Fix invalid domain response handling
Patch-mainline: v4.10-rc1
Git-commit: bf9f31190aa176c43c15cf58b60818d325e0f851
References: bnc#1009718, LTC#149851

Description:  kernel: Fix invalid domain response handling
Symptom:      Crypto adapters switching to offline
              when invalid domain number is used
              on a crypto request.
Problem:      Firmware introduced two new reply codes
              indicating that there is a wrong domain
              number used. These new reply codes are handled
              default in the device driver resulting in
              switching off the adapter. So after processing
              some of these malformed requests all crypto
              adapters are offline. These kind of requests
              can be produced by some wrong implemented
              requests sent out by the TKE.
Solution:     The two new reply codes are handled in a way
              that the calling process gets an errno EINVAL
              without switching the adapter offline.
Reproduction: By operating on the TKE 8.0 console panel
              "Crypto Services" on the "Module Details"
              panel.
Upstream-ID:  bf9f31190aa176c43c15cf58b60818d325e0f851
Problem-ID:   149851

Signed-off-by: Harald Freudenberger <freude@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.com>
---
 drivers/s390/crypto/zcrypt_error.h |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/s390/crypto/zcrypt_error.h
+++ b/drivers/s390/crypto/zcrypt_error.h
@@ -65,6 +65,7 @@ struct error_hdr {
 #define REP82_ERROR_FORMAT_FIELD     0x29
 #define REP82_ERROR_INVALID_COMMAND  0x30
 #define REP82_ERROR_MALFORMED_MSG    0x40
+#define REP82_ERROR_INVALID_DOMAIN_PRECHECK 0x42
 #define REP82_ERROR_RESERVED_FIELDO  0x50	/* old value	*/
 #define REP82_ERROR_WORD_ALIGNMENT   0x60
 #define REP82_ERROR_MESSAGE_LENGTH   0x80
@@ -72,6 +73,7 @@ struct error_hdr {
 #define REP82_ERROR_OPERAND_SIZE     0x84
 #define REP82_ERROR_EVEN_MOD_IN_OPND 0x85
 #define REP82_ERROR_RESERVED_FIELD   0x88
+#define REP82_ERROR_INVALID_DOMAIN_PENDING  0x8A
 #define REP82_ERROR_TRANSPORT_FAIL   0x90
 #define REP82_ERROR_PACKET_TRUNCATED 0xA0
 #define REP82_ERROR_ZERO_BUFFER_LEN  0xB0
@@ -97,6 +99,8 @@ static inline int convert_error(struct z
 	case REP82_ERROR_OPERAND_SIZE:
 	case REP82_ERROR_EVEN_MOD_IN_OPND:
 	case REP88_ERROR_MESSAGE_MALFORMD:
+	case REP82_ERROR_INVALID_DOMAIN_PRECHECK:
+	case REP82_ERROR_INVALID_DOMAIN_PENDING:
 	//   REP88_ERROR_INVALID_KEY		// '82' CEX2A
 	//   REP88_ERROR_OPERAND		// '84' CEX2A
 	//   REP88_ERROR_OPERAND_EVEN_MOD	// '85' CEX2A
